proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+to penalize, even if the application is not CPI'd to.
+
+The checks on the application fees will be taken care of by Solana runtime.
+Total application fees paid will be included in the transaction through a
+special `PayApplicationFees` similar to the `ComputeBudget` program so that it
+is easier to calculate the total amount of fees that will be paid. In case of
+nonpayment or partial payment, the program is never executed. This instruction
+cannot be CPI'ed into.
+
+Application developers will be able to rebate these fees through a new runtime
+mechanism, but only if the transaction actually succeeds, this way CU heavy
+simulation of the program while write-locking the respective accounts will
+require full fee payment.
+
+### Payment
+
+When the cluster receives a new transaction, `PayApplicationFees` instruction is
+decoded to calculate the total fees required for the transaction message. Then
+we verify that the fee-payer has a minimum balance of:
+`per-transaction base fees` + `prioritization fees` +
+`maximum application fees to be paid`
+
+If the payer does not have enough balance, the transaction is not scheduled
+and fails with an error `InsufficientFunds`. Consider this case the same as if
+the payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+application fees. For all accounts with application fees, the fees paid should
+be greater than the required fees. In case of overpayment, the difference is
+stored in a variable and will be paid back to the payer in any case. In case the
+application fees are insufficiently paid or not paid then the transaction fails
+with an error `ApplicationFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The application fees minus rebates and minus overpayment are transferred to
+the respecitve accounts lamport balance, from where the owner can collect them.
+
+### Configuration
+
+Application fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+The ledger will need to store the amount of lamports required for every account.
+Hence changes should be constrained like any account write to the owner of the
+account. All programs will need to implement special instructions so that the
+authority of the accounts can sign to update application fees on the accounts.
+
+Setting the amount to `0` disables application fees. The maximum amount of
+application fees that can be set for an account will be limited to a constant
+amount of SOL (e.g. 20 SOL) so that accounts cannot become inaccessible through
+simple programming errors.
+
+### Rebate
+
+The owner of the account can issue a rebate of the application fees paid to
+write lock a specific account. Similar to the configuration this will need to be
```

Comment:
"the configuration": sorry, what configuration?

Reply:
See `### Configuration` in line 157, the section right before "Rebate"